File: typescript/src/TennisGame1.ts
import {TennisGame} from './TennisGame';
import {Player} from "./Player";

export class TennisGame1 implements TennisGame {
    private player1: Player;
    private player2: Player;

    constructor(player1Name: string, player2Name: string) {
        this.player1 = new Player(player1Name);
        this.player2 = new Player(player2Name);
    }

    wonPoint(playerName: string): void {
        // TODO replace with map?
        let player: Player = this.player1.areYou(playerName) ? this.player1 : this.player2;
        player.winPoint();
    }


    getScore(): string {
        let p1 = this.player1;
        let p2 = this.player2;
        if (p1.points === p2.points) {
            return equalityScore(p1);
        }
        else if (p1.points >= 4 || p2.points >= 4) {
            return this.advantageOrWin(p1, p2);
        }
        else {
            return pointScore(p1, p2);
        }
    }

    private advantageOrWin(p1, p2) {
        const minusResult: number = p1.points - p2.points;
        const leader = minusResult > 0 ? p1 : p2;
        let typeOfScore = Math.abs(minusResult) === 1 ? 'Advantage ' : 'Win for ';
        return typeOfScore + leader.name;
    }
}

function pointScore(p1: Player, p2: Player) {
    return p1.pointTitle() + '-' + p2.pointTitle();
}

function equalityScore(player: Player) {
    if (player.points >= 3) {
        return "Deuce"
    } else {
        return player.pointTitle() + "-All";
    }
}

File: java/src/main/java/TennisGame1.java
import java.util.*;

public class TennisGame1 implements TennisGame {

    private int m_score1 = 0;
    private int m_score2 = 0;
    private String player1Name;
    private String player2Name;

    private List<String> pointScores = Arrays.asList("Love", "Fifteen", "Thirty", "Forty");

    public TennisGame1(String player1Name, String player2Name) {
        this.player1Name = player1Name;
        this.player2Name = player2Name;
    }

    public void wonPoint(String playerName) {
        if (playerName == "player1")
            m_score1 += 1;
        else
            m_score2 += 1;
    }

    public String getScore() {
        String score = "";
        if (hasEqualScoreBeginningOfGame()) {

            score = getPointScore(m_score1) + "-All";

        } else if (isTieBreak()) {

            score = getTieBreakScore();

        } else {

            score = getPointScore(m_score1) + "-" + getPointScore(m_score2);

        }
        return score;
    }

    private String getTieBreakScore() {
        int scoreDiff = Math.abs(m_score1 - m_score2);
        if (scoreDiff == 0) {
             return "Deuce";
        } else if (scoreDiff == 1) {
            return "Advantage " + leadingPlayerName();
        } else {
            return "Win for " + leadingPlayerName();
        }
    }

    private boolean isTieBreak() {
        return m_score1 >= 4 || m_score2 >= 4 || m_score1 == m_score2;
    }

    private boolean hasEqualScoreBeginningOfGame() {
        return m_score1 == m_score2 && m_score1 + m_score1 <= 4;
    }

    private String getPointScore(int m_score1) {
        return this.pointScores.get(m_score1);
    }

    private String leadingPlayerName() {
        return m_score1 > m_score2 ? "player1" : "player2";
    }
}
